File: C#/Toll-calc/Toll-calc/TollCalculator.cs
using System;
using System.Linq;
using Toll_calc.Extensions;
using Toll_calc.Models;
using Toll_calc.Services;

public class TollCalculator
{
    private readonly IHolidayService _holidayService;

    public TollCalculator(IHolidayService holidayService)
    {
        _holidayService = holidayService;
    }

    /**
    * Calculate the total toll fee for one day
    *
    * @param vehicle - the vehicle
    * @param passTimes   - date and time of all passes on one day
    * @return - the total toll fee for that day
    */
    public int GetTollFeeForDay(Vehicle vehicle, DateTime[] passTimes)
    {
        if (null == vehicle)
            throw new ArgumentException("Vehicle can not be null.");
        if (null == passTimes || passTimes.Length == 0)
            return 0;
        if (passTimes.Any(p => p.Date != passTimes[0].Date))
            throw new ArgumentException("All pass times must be from the same date.");

        if (vehicle.IsTollFree() || IsWeekend(passTimes[0]) || _holidayService.IsHoliday(passTimes[0])) return 0;

        var totalFee = 0;
        var orderedPassTimes = passTimes.OrderBy(p => p.TimeOfDay).ToList();
        var intervalStart = orderedPassTimes[0];

        for (var i = 0; i < orderedPassTimes.Count; i++)
        {
            if (orderedPassTimes[i] < intervalStart)
                continue;
            intervalStart = orderedPassTimes[i];
            var endTime = intervalStart.AddHours(1);

            //select the highest fee in interval
            totalFee += (from date in orderedPassTimes
                         where date >= intervalStart && date < endTime
                         select GetTollFee(date))
                .Max();

            intervalStart = endTime;
        }
        return Math.Min(totalFee, 60);
    }

    private bool IsWeekend(DateTime date)
    {
        return date.DayOfWeek == DayOfWeek.Saturday || date.DayOfWeek == DayOfWeek.Sunday;
    }

    private int GetTollFee(DateTime passTime)
    {
        if (passTime.Hour < 6) return (int)Fees.Free;
        if (passTime.FallsBetween(new TimeSpan(6, 0, 0), new TimeSpan(6, 30, 0))) return (int)Fees.Low;
        if (passTime.FallsBetween(new TimeSpan(6, 30, 0), new TimeSpan(7, 0, 0))) return (int)Fees.Medium;
        if (passTime.FallsBetween(new TimeSpan(7, 0, 0), new TimeSpan(8, 0, 0))) return (int)Fees.High;
        if (passTime.FallsBetween(new TimeSpan(8, 0, 0), new TimeSpan(8, 30, 0))) return (int)Fees.Medium;
        //I've assumed the previous condition (where xx:00 - xx:29 in this was free) was incorrect (//else if (hour >= 8 && hour <= 14 && minute >= 30 && minute <= 59) return 8;)
        //I've changed it to Low fee for the entire span.
        if (passTime.FallsBetween(new TimeSpan(8, 30, 0), new TimeSpan(15, 0, 0))) return (int)Fees.Low;
        if (passTime.FallsBetween(new TimeSpan(15, 0, 0), new TimeSpan(15, 30, 0))) return (int)Fees.Medium;
        if (passTime.FallsBetween(new TimeSpan(15, 30, 0), new TimeSpan(17, 0, 0))) return (int)Fees.High;
        if (passTime.FallsBetween(new TimeSpan(17, 0, 0), new TimeSpan(18, 0, 0))) return (int)Fees.Medium;
        if (passTime.FallsBetween(new TimeSpan(18, 0, 0), new TimeSpan(18, 30, 0))) return (int)Fees.Low;
        return (int)Fees.Free;
    }

    [Obsolete]
    public int GetTollFee_Old(DateTime date, Vehicle vehicle)
    {
        int hour = date.Hour;
        int minute = date.Minute;

        if (hour == 6 && minute >= 0 && minute <= 29) return 8;
        else if (hour == 6 && minute >= 30 && minute <= 59) return 13;
        else if (hour == 7 && minute >= 0 && minute <= 59) return 18;
        else if (hour == 8 && minute >= 0 && minute <= 29) return 13;
        else if (hour >= 8 && hour <= 14 && minute >= 30 && minute <= 59) return 8;
        else if (hour == 15 && minute >= 0 && minute <= 29) return 13;
        else if (hour == 15 && minute >= 0 || hour == 16 && minute <= 59) return 18;
        else if (hour == 17 && minute >= 0 && minute <= 59) return 13;
        else if (hour == 18 && minute >= 0 && minute <= 29) return 8;
        else return 0;
    }


    private enum Fees
    {
        Free = 0,
        Low = 8,
        Medium = 13,
        High = 18
    }
}

File: src/main/java/com/evolve/tollcalculator/TollCalculator.java
package com.evolve.tollcalculator;

import java.time.LocalTime;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.DayOfWeek;
import java.util.*;

import java.util.stream.Collectors;

import com.evolve.services.HolidayService;
import com.evolve.services.TollFeeService;
import com.evolve.vehicles.Vehicle;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class TollCalculator {
    private HolidayService holidayService;
    private TollFeeService tollFeeService;
    private static final Logger logger = LogManager.getLogger(TollCalculator.class);
    public static final int MAX_FEE_IN_ONE_DAY = 60;

    public TollCalculator(HolidayService holidayService, TollFeeService tollFeeService) {
        this.holidayService = holidayService;
        this.tollFeeService = tollFeeService;
    }

    /**
     * Calculate the total toll fee for one day
     *
     * @param vehicle - the vehicle
     * @param dates   - date and time of all passes on one day
     * @return - the total toll fee for that day
     */
    public int getTollFee(Vehicle vehicle, Date... dates) {
        if (vehicle == null)
            throw new IllegalArgumentException("Vehicle can't be null.");
        if (dates == null || dates.length == 0)
            return 0;
        if (Arrays.stream(dates).anyMatch(Objects::isNull)) {
            logger.error("The input dates contain null: " + Arrays.toString(dates));
            throw new IllegalArgumentException("The input dates has null entry.");
        }

        // check if all dates are of the same day
        List<LocalDate> temp = Arrays.stream(dates).map(d -> d.toInstant().atZone(ZoneId.systemDefault()).toLocalDate()).collect(Collectors.toList());
        LocalDate entry1 = temp.get(0);
        if (temp.stream().anyMatch(d -> !d.equals(entry1))) {
            logger.error("The input dates are not of the same date: " + Arrays.toString(dates));
            throw new IllegalArgumentException("All dates shall be on the same date.");
        }
        if (isTollFreeVehicle(vehicle) || isTollFreeDate(entry1))
            return 0;

        /*
         * Drop the toll-free entries.
         * Consider the following case:
         * Tim      fee     comment
         * 05:10    0       start of 1st interval if toll-free entries are not dropped
         * 06:05    8       start of 1st interval if toll-free entries are dropped
         * 06:20    8       start of 2nd interval if toll-free entries are not dropped
         *
         * If the toll-free entries are not dropped, the total fees are 8 + 8 = 16.
         * Otherwise it is 8.
         * Thus we should drop toll-free entries so that free is free.
         */
        List<LocalTime> localTimes = Arrays.stream(dates).map(d -> d.toInstant().atZone(ZoneId.systemDefault())
                .toLocalTime()).filter(d -> getTollFee(d) > 0).collect(Collectors.toList());
        if (localTimes.isEmpty())
            return 0;

        // end of one-hour interval
        LocalTime intervalEnd = localTimes.get(0).plusHours(1);
        int totalFee = 0;
        int maxFee = 0;
        for (LocalTime time : localTimes) {
            int tollFee = getTollFee(time);
            if (time.isAfter(intervalEnd)) {
                // passed one-hour period, add highest toll in this hour to total fee.
                totalFee += maxFee;
                intervalEnd = time.plusHours(1);
                maxFee = tollFee;
            } else {
                if (maxFee < tollFee)
                    maxFee = tollFee;
            }
        }
        // add the last entry to total. Note: this entry is not counted within the loop.
        totalFee += maxFee;

        return (Math.min(totalFee, MAX_FEE_IN_ONE_DAY));
    }

    /**
     * Get the toll fee at a specified time.
     *
     * @param time  the specified time
     * @return  the toll fee should be charged at the specified time
     */
    private int getTollFee(LocalTime time) {
        // discard the seconds part
        LocalTime t = LocalTime.of(time.getHour(), time.getMinute());
        return tollFeeService.getTollFee(t);
    }

    /**
     * Check if the given date is a toll-free date.
     * @param date  the specified date
     * @return  true if it is toll-free, false otherwise
     */
    private Boolean isTollFreeDate(LocalDate date) {
        DayOfWeek dayOfWeek = date.getDayOfWeek();
        if (dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY)
            return true;
        return holidayService.isHoliday(date);
    }

    /**
     * Check if a vehicle is toll-free.
     * @param vehicle   the vehicle
     * @return  true if the vehicle is toll-free, false otherwise
     */
    private boolean isTollFreeVehicle(Vehicle vehicle) {
        return vehicle.isTollFree();
    }
}
